### DISCUSSION TITLE
Is it possible to run gunicorn under Fil?

### DISCUSSION THREAD

**Discussion by craigds** [pythonspeed/filprofiler#362] [Q&A] [answered]
Thanks for making Fil. 

We're trying to debug a slow leak in a webserver (resulting in an eventual OOM kill), and at first glance Fil looks ideal. However, I'm having trouble running gunicorn under it. my first effort looks something like this:

```bash
fil-profile -o /fil-profile-output/ --no-browser run \
    gunicorn --threads 4 --workers 1 --worker-class gthread myapp.wsgi -b 0.0.0.0:8080
```

However this doesn't seem to work, due to the way that gunicorn forks worker processes before running our code. To its credit, Fil does give me this warning at startup:

```
=fil-profile= WARNING: Fil does not (yet) support tracking memory in subprocesses.
```

AFAICT to make this work I'd have to effectively get gunicorn to run worker subprocesses via `fil-profile python` instead of merely forking and running the app code. I can't see that happening. Are there any other ways this could work?

Thanks

**Comment by itamarst** [comment, marked as answer]
It's possible BTW that warning is unrelated to gunicorn (e.g. some Python modules run subprocesses on import). However, it does seem like even in `gthread` mode gunicorn launches subprocesses.

One thing you can try is using a different WSGI server, given that presumably the memory leak is in your business logic, not the gunicorn. E.g. uvicorn appears to not use subprocesses if number of workers is 1, or `pip install twisted` and then `PYTHONPATH=. twist web --wsgi=test.app` works for me given `test.py` with an `app()` function.

**Reply by craigds** [reply]
thanks, I suspected that might be the answer. I'll give uvicorn a crack and see how we go.
